airbyte-workers/src/main/java/io/airbyte/workers/process/ExitCodeWatcher.java

```diff
@@ -20,52 +20,101 @@
  * be able to retrieve the exit code).
  */
 @Slf4j
-public class ExitCodeWatcher implements Watcher<Pod> {
+public class ExitCodeWatcher implements ResourceEventHandler<Pod> {
 
+  private final String podName;
+  private final String podNamespace;
   private final Consumer<Integer> onExitCode;
-  private final Consumer<WatcherException> onWatchFailure;
-  private boolean exitCodeRetrieved = false;
+  private final Runnable onWatchFailure;
+  /**
+   * This flag is set to false when we either (a) find the pod's exit code, or (b) when the pod is
+   * deleted. This is so that we call exactly one of onExitCode and onWatchFailure, and we make that
+   * call exactly once.
+   */
+  private boolean active = true;
```

Comment:
do we need to worry about this being thread safe?

Reply:
good question! I'm still working with this against a local instance (very WIP; tests aren't even passing yet) so I'll figure this out while I'm at it.

Reply:
couldn't find a definitive answer, but moved it into an AtomicBoolean just to be safe.